metadata/src/main/java/org/apache/kafka/image/MetadataProvenance.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.image;
+
+import org.apache.kafka.raft.OffsetAndEpoch;
+
+import java.util.Objects;
+
+
+/**
+ * Information about the source of a metadata image.
+ */
+public final class MetadataProvenance {
+    public static final MetadataProvenance EMPTY = new MetadataProvenance(-1L, 0, 0L);
+
+    private final long offset;
+    private final int epoch;
+    private final long lastContainedLogTimeMs;
+
+    public MetadataProvenance(
+        long offset,
+        int epoch,
+        long lastContainedLogTimeMs
+    ) {
+        this.offset = offset;
+        this.epoch = epoch;
+        this.lastContainedLogTimeMs = lastContainedLogTimeMs;
+    }
+
+    public OffsetAndEpoch offsetAndEpoch() {
+        return new OffsetAndEpoch(offset, epoch);
+    }
+
+    public long offset() {
+        return offset;
+    }
+
+    public int epoch() {
+        return epoch;
+    }
+
+    public long lastContainedLogTimeMs() {
+        return lastContainedLogTimeMs;
+    }
+
+    public String snapshotName() {
```

Comment:
Is this used anywhere? Is it meant to match the name of the snapshot file?

Reply:
It is used in the new metadata loader, which is not part of this PR. It is the name that a snapshot with the given provenance (offset, epoch) would have. I will add a Javadoc comment.